aproxy.go

```diff
@@ -0,0 +1,404 @@
+	var closed atomic.Bool
+	go func() {
+		_, err := io.Copy(proxyConn, conn)
+		if err != nil && !closed.Load() {
+			logger.Error("failed to relay network traffic to proxy", "error", err)
+		}
+		closed.Store(true)
+		_ = proxyConn.Close()
+	}()
+	_, err := io.Copy(conn, proxyConn)
+	if err != nil && !closed.Load() {
+		logger.Error("failed to relay network traffic from proxy", "error", err)
+	}
+	closed.Store(true)
+}
+
+func RelayHttp(conn io.ReadWriter, proxyConn io.ReadWriteCloser, logger *slog.Logger) {
+	defer func() {
+		_ = proxyConn.Close()
+	}()
+	req, err := http.ReadRequest(bufio.NewReader(conn))
+	if err != nil {
+		logger.Error("failed to read HTTP request from connection", "error", err)
+		return
+	}
+	req.URL.Host = req.Host
+	req.URL.Scheme = "http"
+	req.Header.Set("Connection", "close")
+	if err := req.WriteProxy(proxyConn); err != nil {
+		logger.Error("failed to send HTTP request to proxy", "error", err)
+		return
+	}
+	resp, err := http.ReadResponse(bufio.NewReader(proxyConn), req)
+	if err != nil {
+		logger.Error("failed to read HTTP response from proxy", "error", err)
+		return
+	}
+	resp.Header.Set("Connection", "close")
+	if err := resp.Write(conn); err != nil {
+		logger.Error("failed to send HTTP response to connection", "error", err)
+		return
+	}
+}
+
+func HandleConn(conn net.Conn, proxy string) {
+	defer func() { _ = conn.Close() }()
+	logger := slog.With("src", conn.RemoteAddr())
+	dst, err := GetOriginalDst(conn.(*net.TCPConn))
+	if err != nil {
+		slog.Error("failed to get connection original destination", "error", err)
+		return
+	}
+	logger = logger.With("original_dst", dst)
+	var host string
+	var relay func(conn io.ReadWriter, proxyConn io.ReadWriteCloser, logger *slog.Logger)
+	var dialProxy func(proxy string) (net.Conn, error)
+	consigned := NewPrereadConn(conn)
+	switch dst.Port {
+	case 443:
+		relay = RelayTcp
+		sni, sniErr := PrereadSNI(consigned)
+		if sniErr != nil {
+			err = sniErr
+		} else {
+			host = fmt.Sprintf("%s:%d", sni, dst.Port)
+			dialProxy = func(proxy string) (net.Conn, error) { return DialProxyConnect(proxy, host) }
+		}
+	case 80:
+		relay = RelayHttp
+		host, err = PrereadHttpHost(consigned)
+		if !strings.Contains(host, ":") {
+			host = fmt.Sprintf("%s:%d", host, dst.Port)
+		}
+		dialProxy = DialProxy
+	default:
+		logger.Error(fmt.Sprintf("unknown destination port: %d", dst.Port))
+		return
+	}
+	if err != nil {
+		logger.Error("failed to preread host from connection", "error", err)
+		return
+	}
+	logger = logger.With("host", host)
+	proxyConn, err := dialProxy(proxy)
+	if err != nil {
+		logger.Error("failed to connect to http proxy", "error", err)
+		return
+	}
+	logger.Info("relay connection to http proxy")
+	relay(consigned, proxyConn, logger)
+}
+
+func main() {
+	proxyFlag := flag.String("proxy", "", "upstream HTTP proxy address in the 'host:port' format")
+	flag.Parse()
+	listenAddr := &net.TCPAddr{IP: net.ParseIP("0.0.0.0"), Port: 8443}
```

Comment:
Should these be specified with an `-addr` command line flag, which would default to `0.0.0.0:8443`, or is hard-coding this fine?

Reply:
No, it's just we don't need this to be configurable, I will add this in the future version :)